Project/Gem/Source/Vision/IdealVisionSystem.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root
+ * of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "IdealVisionSystem.h"
+#include <Atom/RPI.Public/AuxGeom/AuxGeomDraw.h>
+#include <Atom/RPI.Public/AuxGeom/AuxGeomFeatureProcessorInterface.h>
+#include <Atom/RPI.Public/Scene.h>
+#include <AzCore/Component/TransformBus.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/SerializeContext.h>
+#include <AzFramework/Physics/Common/PhysicsSceneQueries.h>
+#include <AzFramework/Physics/PhysicsScene.h>
+#include <AzFramework/Physics/PhysicsSystem.h>
+#include <AzFramework/Physics/Shape.h>
+#include <AzFramework/Physics/SystemBus.h>
+#include <ROS2/ROS2Bus.h>
+#include <ROS2/Sensor/ROS2SensorComponent.h>
+#include <ROS2/Utilities/ROS2Conversions.h>
+#include <ROS2/Utilities/ROS2Names.h>
+
+namespace ROS2::Demo
+{
+    constexpr char Detection2DType[] = "vision_msgs::msg::Detection2D";
+    constexpr char Detection3DType[] = "vision_msgs::msg::Detection3D";
+    constexpr char DetectionPoseArrayType[] = "geometry_msgs::msg::PoseArray";
+
+    IdealVisionSystem::IdealVisionSystem()
+    {
+        TopicConfiguration tc;
+
+        tc.m_type = DetectionPoseArrayType;
+        tc.m_topic = "detectionsPoseArray";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        tc.m_type = Detection2DType;
+        tc.m_topic = "detections2D";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        tc.m_type = Detection3DType;
+        tc.m_topic = "detections3D";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        m_sensorConfiguration.m_frequency = 10;
+    }
+
+    AZ::Vector3 GetPointIn3D(const AZ::Vector2& point2D, float depth, AZ::Matrix3x3 cameraMatrix)
+    {
+        cameraMatrix = cameraMatrix.GetInverseFull();
+        AZ::Vector3 point3D;
+        point3D.SetX(point2D.GetX());
+        point3D.SetY(point2D.GetY());
+        point3D.SetZ(1.0f);
+        point3D = depth * cameraMatrix * point3D;
+        return point3D;
+    }
+
+    AZ::Vector2 GetPointIn2D(const AZ::Vector3& point3D, const AZ::Matrix3x3& cameraMatrix)
+    {
+        AZ::Vector3 point2D;
+        point2D = cameraMatrix * point3D;
+        point2D = point2D / point2D.GetZ();
+        return AZ::Vector2(aznumeric_cast<float>(point2D.GetX()), aznumeric_cast<float>(point2D.GetY()));
+    }
+
+    void IdealVisionSystem::Activate()
+    {
+        constexpr float MaximalDetection = 2.0f;
```

Comment:
What is the meaning of this constant? It is passed as depth parameter. Please use a clearer name - is it about detection frustum scaled through its height?